packages/flutter/lib/src/widgets/selection_gestures.dart

```diff
@@ -0,0 +1,1251 @@
+
+  void _acceptDrag(PointerMoveEvent event) {
+    _checkTapCancel();
+    if (dragStartBehavior == DragStartBehavior.start) {
+      _initialPosition = _initialPosition + OffsetPair(global: event.delta, local: event.localDelta);
+    }
+    _checkStart(event);
+    if (event.localDelta != Offset.zero) {
+      final Matrix4? localToGlobal = event.transform != null ? Matrix4.tryInvert(event.transform!) : null;
+      final Offset correctedLocalPosition = _initialPosition.local + event.localDelta;
+      final Offset globalUpdateDelta = PointerEvent.transformDeltaViaPositions(
+        untransformedEndPosition: correctedLocalPosition,
+        untransformedDelta: event.localDelta,
+        transform: localToGlobal,
+      );
+      final OffsetPair updateDelta = OffsetPair(local: event.localDelta, global: globalUpdateDelta);
+      _correctedPosition = _initialPosition + updateDelta; // Only adds delta for down behaviour
+      _checkUpdate(event);
+      _correctedPosition = null;
+    }
+  }
+
+  void _checkDrag(PointerMoveEvent event) {
+    final Matrix4? localToGlobalTransform = event.transform == null ? null : Matrix4.tryInvert(event.transform!);
+    _globalDistanceMoved += PointerEvent.transformDeltaViaPositions(
+      transform: localToGlobalTransform,
+      untransformedDelta: event.localDelta,
+      untransformedEndPosition: event.localPosition
+    ).distance * 1.sign;
+    if (_hasSufficientGlobalDistanceToAccept(event.kind, gestureSettings?.touchSlop)) {
+      if (event.buttons == kSecondaryButton) {
+        // Reject a right click drag. It is possible that the recognizer may have
+        // already won the arena at this point, if it did then we should clear some
+        // state to prepare to track the next [PointerDownEvent].
+        if (_wonArenaForPrimaryPointer) {
+          _stopDeadlineTimer();
+          _giveUpPointer(event.pointer);
+          _resetTaps();
+          _resetDragUpdateThrottle();
+          _initialButtons = null;
+        }
+        resolve(GestureDisposition.rejected);
+        return;
+      }
+      _start = event;
+      _dragState = _DragState.accepted;
+      resolve(GestureDisposition.accepted);
+    }
+  }
+
+  void _checkTapDown(PointerDownEvent event) {
+    if (_sentTapDown) {
+      return;
+    }
+
+    final TapDragDownDetails details = TapDragDownDetails(
+      globalPosition: event.position,
+      localPosition: event.localPosition,
+      kind: getKindForPointer(event.pointer),
+      consecutiveTapCount: consecutiveTapCount,
+      keysPressedOnDown: keysPressedOnDown,
+    );
+
+    switch (_initialButtons) {
+      case kPrimaryButton:
+        if (onTapDown != null) {
+          invokeCallback('onTapDown', () => onTapDown!(details));
+        }
+        break;
+      case kSecondaryButton:
+        if (onSecondaryTapDown != null) {
+          invokeCallback('onSecondaryTapDown', () => onSecondaryTapDown!(details));
+        }
+        break;
+      default:
```

Comment:
Yes because we should start tracking `_initialButtons` in `addAllowedPointer`. If `_initialButtons` is null at this point we cannot determine which callback to fire.

Reply:
I added a few asserts here in case.